dl/src/main/scala/com/intel/analytics/bigdl/nn/Recurrent.scala

```diff
@@ -45,19 +98,26 @@ class Recurrent[T : ClassTag] (
     module = modules(0)
     transform = modules(1)
 
-    batchSize = input.size(1)
-    times = input.size(2)
+    batchSize = input.size(batchDim)
+    times = input.size(timeDim)
+
+    hidden.resize(Array(times + 1, batchSize, hiddenSize))
+
+    if (timeDim == 2 && !legacyMode) {
+      fInput = input.transpose(batchDim, timeDim).contiguous
+    } else {
+      fInput = input
+    }
 
-    output.resize(Array(batchSize, times, hiddenSize))
-    hidden.resize(Array(batchSize, times + 1, hiddenSize))
+    output.resize(Array(input.size(1), input.size(2), hiddenSize))
 
     var i = 1
     while (i <= times) {
-      val curInput = T(input.select(2, i), hidden.select(2, i))
+      val curInput = T(fInput.select(dataSelect, i), hidden.select(hiddenSelect, i))
       val currentOutput = module.updateOutput(curInput)
       transform.updateOutput(currentOutput)
-      output.select(2, i).copy(transform.output.toTensor)
-      hidden.select(2, i + 1).copy(transform.output.toTensor)
+      output.select(outputSelect, i).copy(transform.output.toTensor)
+      hidden.select(hiddenSelect, i + 1).copy(transform.output.toTensor)
```

Comment:
Maybe we can change hidden to Array[Tensor](times + 1), and then we can avoid the copy here: hidden(i+1) = transform.output